### ISSUE TITLE
PySide2 QAbstractListmodel rowCount not implemented?

### ISSUE THREAD

**Issue by shpi** [Nuitka/Nuitka#1009]
nuitka version > 0.6.12  (PySide2 QT 5.15.1)

minimal example: https://github.com/xtompok/PySide2_QML_tutorial/blob/master/04_city_list/city_list/city_list.py

It seems QAbstractListModel rowCount doesn't work.( Anymore? )

It throws "NotImplementedError: pure virtual method QAbstractListModel.rowCount() not implemented" error.

**Comment by kayhayen** [comment]
It seems there is a 5.15.2 available, which I will be used now to reproduce the issue, and hopefully fail to, but I think the version number of PySide2 you are using should be the one that had our patches merged.

**Comment by shpi** [comment]
Yes 

>>> print(PySide2.__version__)
5.15.2

Same Issue. Thanks

**Comment by kayhayen** [comment]
I noticed that there is of course zero support for PySide2 with Nuitka standalone, and have at first added a new plugin that shares code with the PyQt4/5 one, and managed to adapt for QML differences, etc. and get it to work, and then I reproduced your issue now. 

I am separating the plugins, because it's getting too unreadable. I guess you are aware of that you probably want to use --standalone or do you not care about that?

**Comment by shpi** [comment]
We don't plan to use standalone yet, because this would complicate OTA with github.

**Comment by kayhayen** [comment]
@shpi ah, that's OK... however we need a plugin anyway, so no harm there

**Comment by shpi** [comment]
I've found a similar issue in the qt bug tracker:

https://bugreports.qt.io/browse/PYSIDE-829

It seem that function overloading does not work properly with shiboken2 ?

**Comment by kayhayen** [comment]
That is the patches I am referring to, the issue is not closed though indeed, but mostly because I have never confirmed this to be complete, and obviously it is not. I am devising a workaround for PySide2 while working on a new patch that hopefully will allow to have a better PySide2, but since you have priority with this issue, that's where I am going first, aiming at a class decorator that avoids the issue that triggers PySide2.

**Comment by kayhayen** [comment]
I am assuming the pure virtual thing for added later. It breaks easy workaround, but I devised one, and got it working just now, which wraps functions in uncompiled functions, so the pure virtual overload check works. In our patches, we covered other code, but obviously no version of PySide2 has worked so far it seems, at least not if something has pure functions like these models do.

I shall try and enhance the wrapping, so it's only done on compiled functions now, and then only for things that are overloads, so it doesn't unnecessarily wrap too many things.

Which Python version are you using @shpi ? The reason I am asking, is that there are subclassing hooks that might be put to use here in newer Python versions, which would allow to "decorate" these classes purely based on the plugin decision and module post-load code installed for PySide2 with the big advantage to not have to modify any source code at all, but for Python2 that won't be available, and I actually forgot which Python3 does it, but you get the idea.

**Comment by shpi** [comment]
We use python 3.9, so this should be no issue

**Comment by kayhayen** [comment]
I checked and PEP 487 which provides `__init_subclass__` is Python 3.6 or higher. Since it will need a new PySide2 anyway (I don't think we want to get earlier PySide2 versions to work), I doubt that Python2 support will be a limiting factor, and if we find it to be, other workarounds can be devised.

I will now check how legal it is to override these attributes from the outside, if that works it will be very easy. If that doesn't work, I think making a wrapper class that overloads it, and replace the name in PySide2 exported things, ought to work instead.

**Comment by kayhayen** [comment]
Yup, overwriting it just works. Strangely, the "old" function there, from `shiboken` which is at the core of the bindings of PySide too, refuses to be called with arguments, whereas Python calls it with the class an argument.

```
Traceback (most recent call last):
  File "C:\Users\kayha\repos\Py2C\MiniPyside2.py", line 70, in <module>
    class CityListModel(QAbstractListModel):
  File "C:\Users\kayha\repos\Py2C\MiniPyside2.py", line 61, in my_init_subclass
    return old_init_subclass(nuitka_wrap(cls))
TypeError: __init_subclass__() takes no arguments (1 given)
```

I am monkey patching the very base of these models, as they probably all share the issue. So now I am going to integrate that into the PySide2 plugin, so it requires no source modification, and I expect that to issue to workarounded soon. Then I shall take a peek look at shiboken there, what is causing the error message, likely an incompatible type check. I noticed that it also rejects partial functions from functools e.g. which is not nice either, so it's a general issue of PySide2 really, and they should accept callables, much like our other patches made other places do.

**Comment by kayhayen** [comment]
I just pushed to the factory branch an update which will make it work with `--enable-plugin=pyside`, this plugin will become automatically active one validated, it doesn't detect the need to use it. This enforces a recursing into `PySide2` top level module, so it can attach the post load code to it. 

For accelerated mode, mixing of modules has been seen as a problem in that past, so it may happen that it fails to load Python child modules. If that happens to you, let me know. You can workaround that with `--include-module` on the names, but I would like to be aware, and do it in the plugin or fix the loading by supporting that in Nuitka (so far only extension modules can be loaded from disk once a package is compiled). Should this prove problematic, the plugin could also enforce what `--include-package=PySide2` will do, but i would hope for it to not be necessary. Obviously, for standalone, that's all done anyway.

The monkey patching only avoids `__init__.py` at this time. I would like you to confirm that the example of working. I am now going to exclude all dunder methods and attempt to limit the monkey patch of classes to only ones which are actually relevant.

Should you find other issues with PySide2, please consider making a new issue ouf of those. For mixing issues as described above, this one will do, it's been reported a couple of times already, so anything new would be a duplicate.

Find the current factory code here: https://nuitka.net/doc/factory.html

**Comment by shpi** [comment]
wow, got it working on my notebook and on one shpi :-)  really nice. big thanks. i will test if cross compiling works also generated c files. compiling on a rpi zero takes a bit time ;-)

**Comment by kayhayen** [comment]
So, cross compilation is not supported, some people are doing it, but it's definitely not recommended, the design is currently not lending itself to it. I would recommend you to run a virtual machine on a bigger box to do it, also with more memory, then ARM will do.

But to give you an update, I made all the necessary enhancements to the workaround, now it limits itself to only doing the necessary work, where before it potentially modified a lot of code, that's not even in your realms. It still works, and it's still on factory, but I expect that to change within ours. Also a stable release with this is immanent.

The PySide2 plugin got cleaned up a lot, and now shares a base class with PyQt5, making it way more clean. I think I can maintain that. One thing I do notice however, is that the "qml" handling for standalone is still something to be provided by the user. In accelerated mode, you are not affected by lack of plugins, but in standalone with default arguments, your code shows an empty window, due to missing included QML files for QtQuick.

I think, I want to move this out, but that's probably not happening before release, as getting 0.6.13 out is the priority right now, and you don't need that now. I just like it when things have a better out of the box experience. Speaking of that. Plugins for PyQt5, maybe PyQt6 one day, and PySide2 are somewhat conflicting, or should be. That makes it hard to get to a point where they can be abled by default. But that's not possible for this new code anyway right now.

**Comment by kayhayen** [comment]
A further improvement is on factory, turns out that method overloads with too few arguments were not working this way. It will also improve the performance of the workaround, as we now provide a plain argument call with the values. Should there be keyword only arguments, that wouldn't work, but I doubt Qt has those.

But this change is now on develop and adds also standalone support for PySide2.

**Comment by shpi** [comment]
I think one change implemented a bug:

File "/home/dell/.local/lib/python3.8/site-packages/PySide2/__init__.py-postLoad", line 69, in my_init_subclass
  File "/home/dell/.local/lib/python3.8/site-packages/PySide2/__init__.py-postLoad", line 55, in nuitka_wrap
  File "<string>", line 2
    def _wrapped_function_rowCount_1(self, parent=<PySide2.QtCore.QModelIndex(-1,-1,0x0,QObject(0x0)) at 0x7f2cb0916f00>):
                                                  ^
SyntaxError: invalid syntax

**Comment by kayhayen** [comment]
Yep, default values are of course not always that easy to access. Luckily there is `__defaults__` to copy them from, so I will be able to fix this up. Sorry for the breakage.

**Comment by kayhayen** [comment]
I pushed a fix to factory. Basically I changed the code to use signature objects, so values not present are retained, star arguments confused Qt to call it with more or less values, depending on the signature. In your example that would mean that e.g. `parent = ... ` could be removed, as then shiboken will simply not call it with that argument. That's pretty non-standard, but we need to deal with it as it's part of the API.

Therefore now we replicate argument signatures there exactly, e.g. `self, parent = ...` with the annotations removed. But the default values escaped my attention as all code examples have things with nice repr like a constant. These now also get deleted, but the wrapper gets `__defaults__` copied and that makes it work nicely. I hope this should now be perfect.

**Comment by kayhayen** [comment]
Part of the 0.6.13 release.

**Comment by kayhayen** [comment]
Just to add, once https://bugreports.qt.io/browse/PYSIDE-1523 is merged, the workarounds will no more be needed.

**Comment by shpi** [comment]
As far as I understand, there will be no more open releases for PySide2. So good for PySide6 :-) The last open source version was QT5.15.2  The following versions are only for business users. Thx for your effort!
